### PR TITLE
LOG4J2-3116 Add GCP logging layout

### PR DESCRIPTION
**Pull request by vy**
This PR aims to supersede #543.

### CODE REVIEW

File: log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+            // Verify thread name.
+            assertThat(accessor.getString("_thread"))
+                    .isEqualTo(logEvent.getThreadName());
+
+            // Verify logger name.
+            assertThat(accessor.getString("_logger"))
+                    .isEqualTo(logEvent.getLoggerName());
+
+        });
```
Comment (rocketraman) [inline log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java:199]: Much better than my version of the test, thanks. I think I got off on the wrong foot by using Ecs and Gelf layout tests as a starting point.

**Review by rocketraman — COMMENTED** [review]
Couple of minor comments @vy , thanks.

File: log4j-layout-template-json/src/main/resources/GcpLayout.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  },
+  "severity": {
+    "$resolver": "pattern",
+    "pattern": "%level{WARN=WARNING, TRACE=DEBUG, FATAL=EMERGENCY}",
+    "stackTraceEnabled": false
+  },
+  "message": {
+    "$resolver": "pattern",
+    "pattern": "%m"
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json/src/main/resources/GcpLayout.json:17]: Good catch! Will do.

File: log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            }
+            assertThat(accessor.getString("severity")).isEqualTo(expectedSeverity);
+
+            // Verify message.
+            final String expectedMessage = logEvent.getMessage().getFormattedMessage();
+            assertThat(accessor.getString("message")).contains(expectedMessage);
+            final Throwable exception = logEvent.getThrown();
+            if (exception != null) {
+                final String expectedExceptionMessage = exception.getLocalizedMessage();
+                assertThat(accessor.getString("message")).contains(expectedExceptionMessage);
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java:108]: Was working on this and... It has turned out to be pretty difficult. `PatternLayout`s `ExtendedThrowablePatternConverter` uses `ThrowableProxy`, which has a pretty different way of working than `JsonTemplateLayout`s `StackTraceStringResolver`. In order to make `message` and `_exception.stackTrace` match, I needed to make the latter use `%xEx` pattern too. Yet, I still could not succeed in reproducing the `PatternLayout`-serialized stack trace in tests. I will continue working on this tomorrow.

[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java]: Settled down to a very ugly solution:
```java
final String actualMessage = accessor.getString("message");
assertThat(actualMessage)
        .contains(logEvent.getMessage().getFormattedMessage())
        .contains(exception.getLocalizedMessage())
        .contains("at org.apache.logging.log4j.layout.template.json")
        .contains("at java.util.stream.ForEachOps")
        .contains("at org.junit.platform.engine");
```
`ThrowablePatternConverterTest` doesn't do a terrific job either. Hence, should be okay.

**Comment by vy** [comment]
@rocketraman, thanks so much again. I will do necessary git magic tomorrow to make your commits in #543 visible.

**Pull request closed without merging**